### ISSUE TITLE
user address space writes

### ISSUE THREAD

**Issue by brendangregg** [bpftrace/bpftrace#28]
Perhaps we should have a way to call BPF_FUNC_probe_write_user()

https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=96ae52279594470622ff0585621a13e96b700600

This could either be a call, eg, `write()`, or it could just happen if you write to a user address. As an example of the latter, imagine if this worked (needs #4):

```
./src/bpftrace -e 'uprobe:/bin/bash:readline { *uaddr("ps1_prompt") = "BPF-says-hi> "; }'
```

So your bash prompt becomes "BPF-says-hi> " the text time you hit enter.

**Comment by brendangregg** [comment]
I wonder if an explicit call to probe_write_user() might be desirable, to make it clear that the tool can be destructive.

```
write_user(void *dst, void *src, int len)
OR
uwrite(void *dst, void *src, int len)
```

write_user() is named after the BPF call. I think it may fit better with other bpftrace builtins/calls if it was called uwrite().

**Comment by Delta-in-hub** [comment]
We need this feature.

**Comment by jordalgo** [comment]
@Delta-in-hub Mind describing your use case for this feature?

**Comment by Delta-in-hub** [comment]
Hello @jordalgo 
As far as i know, bpftrace now doesn't have the function like `probe_write_user` , which severely limits the use of bpftrace for more complex purposes like doing hotfix for user programe or something like https://github.com/pathtofile/bad-bpf.

bpftrace is a wonderful tool with a very clean syntax. 
But I'm curious why something like `probe_write_user` was intentionally not provided?  If this is to prevent bpftrace from being used for malicious purposes, I don't find it convincing; after all, there is the `system()` function, which can actually do anything.

**Comment by ajor** [comment]
Agreed that it would be a useful feature to have. It's just a lack of time/desire that's meant that it hasn't been implemented yet.

We'd lock this feature behind bpftrace's "unsafe mode", which would help prevent accidental use and alleviate security risks from running other people's unreviewed scripts. Since bpftrace can only be run as root, users are assumed to be trusted and we won't try to prevent anyone from deliberately using a feature like this.

**Comment by Delta-in-hub** [comment]
@ajor I got that.Thank you guys for maintaining bpftrace!

**Comment by jordalgo** [comment]
Closing this issue in favor of:
- https://github.com/bpftrace/bpftrace/issues/3742
